### PR TITLE
render 3D scene on demand

### LINKED ISSUE #509: Geometry preview - on-demand rendering
Enhancement Request
--------------

Makes 3D rendering of geometry more CPU-efficient by using on-demand rendering and not continuous animation.

## Detailed Description

The main implication will be on the fluidity of scene rendering upon user interactions: with continuous animation there's no `render()` call on each change, so it may appear fluid. On the other hand continuous animation eats up the CPU, and can make the whole experience jerky (possibly related with https://github.com/openstudiocoalition/OpenStudioApplication/issues/118)

## Possible Implementation

The change consists in disabling continuous animation and rendering only when something changes in the scene (ie user changing camera position, or clicking an option in the menu). Some testing required on several devices to ensure it doesn't hurt user experience.

### PR DESCRIPTION
**Pull request by antoine-galataud**
Fixes #509

### CODE REVIEW

**Comment by jmarrec** [comment]
@antoine-galataud Thanks for the PR. Could you make a a small `peek` gif showing before/after to ease in reviewing please?

**Comment by antoine-galataud** [comment]
@jmarrec I had prepared a peek capture for this, but actually there's actually no observable difference before and after the change. If you're still interested I can produce one that also includes CPU metrics (ie htop summary in a corner)

**Comment by antoine-galataud** [comment]
Here are 2 screen captures (large model). It's actually slower and less fluid before the change, due to the constant load of animation + recording of the video. Notice also the large diff in CPU load at the beginning (when no user interaction happens).

Before:

before_animate

After:
after_no_animate

**Comment by jmarrec** [comment]
tested locally, did not notice any difference. @macumber wanna take a pass before I merge?

File: src/openstudio_lib/library/geometry_preview.html [hunk 1]
```diff
@@ -2538,6 +2540,7 @@
     } else {
       removeSelection();
     }
+    requestRenderIfNotRequested();
```
Comment (macumber) [inline src/openstudio_lib/library/geometry_preview.html:2543]: Do we need to do this on changing render mode, hiding surfaces, or other things too?

**Comment by macumber** [comment]
Left one comment, I haven't had time to build and try on my own yet.  My other comment is I'd like to make sure all of these improvements make it back to https://github.com/NREL/openstudio-common-measures-gem/ as well.  That way we don't diverge from those measures too much.

Thanks again @antoine-galataud !

File: src/openstudio_lib/library/geometry_preview.html (hunk 1 above)
Comment (antoine-galataud) [reply src/openstudio_lib/library/geometry_preview.html:2543]: Interactions on the dat.gui menu trigger an `onChange` event that is captured by `update()`. I added a call to `requestRenderIfNotRequested` there too (see https://github.com/openstudiocoalition/OpenStudioApplication/pull/510/files#diff-574d82d3d9fea756c6e5daf9d279da971c8ad80e9171c4f1ff33de92938dfbfcR3070).

I think it covers the needs, but maybe you think of other interactions that should request rendering?

**Comment by antoine-galataud** [comment]
@macumber @jmarrec I can handle adding #510 and #497 to https://github.com/NREL/openstudio-common-measures-gem/ once they are both validated and merged here. Just let me know.

File: src/openstudio_lib/library/geometry_preview.html (hunk 1 above)
Comment (macumber) [reply src/openstudio_lib/library/geometry_preview.html:2543]: Gotcha!

**Review by macumber — APPROVED** [review]

**Comment by jmarrec** [comment]
@antoine-galataud That'd be great if you could upstream to common measure gem! Thanks!

**Pull request merged**